community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/pipes/NodeIndexSeeker.scala

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime.interpreted.pipes
+
+import org.neo4j.cypher.internal.frontend.v3_4.helpers.SeqCombiner.combine
+import org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.{Expression, InequalitySeekRangeExpression, PointDistanceSeekRangeExpression, PrefixSeekRangeExpression}
+import org.neo4j.cypher.internal.runtime.interpreted.{ExecutionContext, IsList, makeValueNeoSafe}
+import org.neo4j.cypher.internal.util.v3_4.{CypherTypeException, InternalException}
+import org.neo4j.cypher.internal.v3_4.logical.plans._
+import org.neo4j.internal.kernel.api.{IndexQuery, IndexReference}
+import org.neo4j.values.AnyValue
+import org.neo4j.values.storable._
+import org.neo4j.values.virtual.NodeValue
+
+/**
+  * Mixin trait with functionality for executing logical index queries.
+  *
+  * This trait maps the logical IndexSeekMode and QueryExpression into the kernel IndexQuery classes, which
+  * are passed to the QueryContext for executing the index seek.
+  */
+trait NodeIndexSeeker {
+
+  // dependencies
+
+  def indexMode: IndexSeekMode
+  def valueExpr: QueryExpression[Expression]
+  def propertyIds: Array[Int]
+
+  // index seek
+
+  protected def indexSeek(state: QueryState,
+                          indexReference: IndexReference,
+                          baseContext: ExecutionContext): Iterator[NodeValue] =
+    indexMode match {
+      case _: ExactSeek =>
+        val indexQueries = computeIndexQueries(state, baseContext)
+        indexQueries.flatMap(query => state.query.indexSeek(indexReference, query)).toIterator
+
+      case LockingUniqueIndexSeek =>
+        val indexQueries = computeExactQueries(state, baseContext)
+        indexQueries.flatMap(indexQuery => state.query.lockingUniqueIndexSeek(indexReference, indexQuery)).toIterator
+
+      case _: SeekByRange =>
+        val indexQueries = computeIndexQueries(state, baseContext)
+        indexQueries.flatMap(query => state.query.indexSeek(indexReference, query)).toIterator
```

Comment:
Well I agree that they are the same, but there is no fallthrough in scala, so that won't work. I guess we could just split it up into `LockingUniqueIndexSeek` and everything else. Will fix!